### PR TITLE
refactor(cat-voices): registration dialog

### PR DESCRIPTION
**Pull request by damian-molinski**
# Description

- Refactor of data passing down the tree inside `RegistrationDialog`.
- Introduce `BlocSelector` which is helpful when working with nested state data objects.
- `Optional.dataOr` extension method.
- Simplify `RegistrationState` with single final class instead of sealed one.
- Rebuilding only widgets when data that they care about changes.

## Related Issue(s)

Refers to #816 

## Please confirm the following checks

* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my code
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] New and existing unit tests pass locally with my changes
* [x] Any dependent changes have been merged and published in downstream module

### CODE REVIEW

File: catalyst_voices/packages/catalyst_voices_blocs/lib/src/bloc_builder_selector.dart [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import 'package:flutter_bloc/flutter_bloc.dart';
+
+typedef BlocValueSelector<T, S> = T Function(S state);
+
+/// Simple extension on [BlocBuilder] but makes it easy to work with nested
+/// objects when you care only about them.
+class BlocBuilderSelector<B extends StateStreamable<S>, S, T>
```
Comment (dtscalac) [inline catalyst_voices/packages/catalyst_voices_blocs/lib/src/bloc_builder_selector.dart:7]: See BlocSelector from flutter_bloc package

Comment (dtscalac) [reply catalyst_voices/packages/catalyst_voices_blocs/lib/src/bloc_builder_selector.dart:7]: I think it does exactly what you need 🗡️

Comment (damian-molinski) [reply catalyst_voices/packages/catalyst_voices_blocs/lib/src/bloc_builder_selector.dart]: Oh! Thats convenient!

File: catalyst_voices/lib/pages/registration/create_keychain/stage/seed_phrase_check_panel.dart [hunk 2]
```diff
@@ -85,40 +41,72 @@ class _SeedPhraseCheckPanelState extends State<SeedPhraseCheckPanel> {
...
+          bool isResetWordsEnabled,
+        )>(
+      selector: (state) => (
+        state.shuffledWords,
+        state.userWords,
+        state.isResetWordsEnabled,
+      ),
+      builder: (context, state) {
+        return _SeedPhraseWords(
+          words: state.$1,
```
Comment (dtscalac) [inline catalyst_voices/lib/pages/registration/create_keychain/stage/seed_phrase_check_panel.dart:101]: Is it possible to use named params for the record? There are two params that have the same type (List<String>), quite easy to do a mistake

**Comment by dtscalac** [comment]
@damian-molinski please update issue description, i.e. BlocBuilderSelector won't be introduced

File: catalyst_voices/lib/pages/registration/registration_info_panel.dart [hunk 3]
```diff
@@ -29,6 +32,7 @@ class RegistrationInfoPanel extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
     return BlocBuilder<RegistrationCubit, RegistrationState>(
+      buildWhen: (previous, current) => previous.step != current.step,
```
Comment (dtscalac) [inline catalyst_voices/lib/pages/registration/registration_info_panel.dart:35]: Would it be possible to use a BlocSelector with the record that you need?

**Review by dtscalac — APPROVED** [review]
LGTM

**Pull request merged**